internal/cli/job_list.go

```diff
@@ -42,6 +45,37 @@ func (c *JobListCommand) Run(args []string) int {
 
 	jobs := resp.Jobs
 
+	// filter jobs out of list on request
+	var tmpj []*pb.Job
+	for _, job := range jobs {
+		if c.flagWorkspace != "" && c.flagWorkspace != job.Workspace.Workspace {
+			continue
+		}
+
+		if c.flagProject != "" && c.flagProject != job.Application.Project {
+			continue
+		}
+
+		if c.flagApp != "" && c.flagApp != job.Application.Application {
+			continue
+		}
+
+		if len(c.flagState) > 0 {
+			found := false
+			for _, state := range c.flagState {
+				if pb.Job_State_name[int32(job.State)] == strings.ToUpper(state) {
+					found = true
+					break
+				}
+			}
+			if !found {
+				continue
+			}
+		}
+		tmpj = append(tmpj, job)
```

Comment:
[server.proto](https://github.com/hashicorp/waypoint/blob/4ef4746c00a0531163c9f55bc60a5dd531b4d93f/pkg/server/proto/server.proto#L193-L195) mentions List Jobs is internal and not for public use yet. I assume that's more of a stability thing rather than just trying to keep it internal, so based on that would it be more appropriate here to modify [ListJobsRequest](https://github.com/hashicorp/waypoint/blob/4ef4746c00a0531163c9f55bc60a5dd531b4d93f/pkg/server/proto/server.proto#L2070) to contain ref fields for filtering by workspace, project, or application while we're here?

Reply:
Hmm, yes, I believe that was a stability choice early on in Waypoint! Maybe we should flip it to just be `ListJobs` instead of `XListJobs` then?

That makes sense to update this endpoint to accept filter requests 👍🏻  If we do filtering on the server, we should implement the filtering in the db state package and update `ListJobRequest` API to take workspace, project, app, job operation, and state `Ref`s to pass on through to the job list db query. That'll be a lot more work than this but that's probably the right way to do it 🤔

Reply:
Yup, server side filtering is the right way to implement this. That's where we're doing other filtering too, so I don't think it would be too much more work!

Reply:
Done.